### ISSUE TITLE
"docker exec wc -c" test fails with Node.js v16.13.2

### ISSUE THREAD

**Issue by jwhitlock** [taskcluster/docker-exec-websocket-server#34]
This test takes 45ms on Node.js v14.17.5, but fails with a timeout on Node.js v16.13.2:

https://github.com/taskcluster/docker-exec-websocket-server/blob/b6ad23c13952e862210ce4659c70e819ef0d19e1/test/test.js#L124

Test output:

```
$ ./test/runtests.sh
#!/bin/bash -ve
# USAGE: Run this file using `npm test` (must run from repository root)

docker run --rm  --name testserver ubuntu sleep 60 &
trap "docker kill testserver;" EXIT; sleep 1;

mocha --exit test/test.js


  trying client
    1) docker exec wc -c
    ✓ cat on server (511ms)
    ✓ exit code (503ms)
    ✓ connection limit
    ✓ automatic pausing (1007ms)
    ✓ session count
    ✓ resize (3514ms)


  6 passing (49s)
  1 failing

  1) trying client
       docker exec wc -c:
     Error: Timeout of 40000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/Users/john/src/docker-exec-websocket-server/test/test.js)
      at listOnTimeout (node:internal/timers:557:17)
      at processTimers (node:internal/timers:500:7)



docker kill testserver;
testserver
error Command failed with exit code 1.
```

This may be the root cause of https://github.com/taskcluster/docker-exec-websocket-client/issues/1.

**Comment by jwhitlock** [comment]
I've tested that the code passes with Node.js 15.14.0 (the last v15 release) and fails with Node.js v16.0.0. So, the breaking change is somewhere in the [changelog](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V16.md#2021-04-20-version-1600-current-bethgriggs)...

When running with ``DEBUG=* mocha test/test.js -f wc`` (all debug messages, and just the one test), I can see where it fails

<summary>A successful run, 15.14.0:</summary>
<details>
<pre>
  docker-exec-server websocket server created for path: "/a" +0ms
  docker-exec-server websocket server created for path: "/a" +1ms
  docker-exec-websocket-server:lib:client ws://localhost:60171/a?tty=false&command=wc&command=-c +0ms
  docker-exec-server connection received +5ms
  modem Sending: {
  path: '/containers/c991255b41376381b9a42ea939019d647aaea2d35b6da73422c3e2854abb57ce/exec',
  method: 'POST',
  headers: { 'Content-Type': 'application/json', 'Content-Length': 105 },
  key: undefined,
  cert: undefined,
  ca: undefined,
  socketPath: '/var/run/docker.sock'
} +7ms
  docker-exec-server 1 sessions created +0ms
  docker-exec-websocket-server:lib:client socket opened +5ms
  docker-exec-websocket-server:lib:client client connected +0ms
  modem Received: {"Id":"99097b8eeb63248a24956714c9cae2694ed08fde0f6bfdf4a092f60c59757f00"}
 +15ms
  modem Sending: {
  path: '/exec/99097b8eeb63248a24956714c9cae2694ed08fde0f6bfdf4a092f60c59757f00/start',
  method: 'POST',
  headers: { 'Content-Type': 'application/json', 'Content-Length': 56 },
  key: undefined,
  cert: undefined,
  ca: undefined,
  socketPath: '/var/run/docker.sock'
} +1ms
  docker-exec-server:data <Buffer 64> +0ms
  docker-exec-server server executed session +30ms
  docker-exec-websocket-server:lib:rcv <Buffer 64> +0ms
  docker-exec-websocket-server:lib:client resumed +30ms
  docker-exec-websocket-server:lib:client 4 +0ms
  docker-exec-websocket-server:lib:client resumed +0ms
  docker-exec-websocket-server:lib:client 5 +0ms
  docker-exec-websocket-server:lib:client resumed +0ms
  docker-exec-websocket-server:lib:client 1 +0ms
  docker-exec-websocket-server:lib:client 0 +0ms
  docker-exec-server got execStream.on('end') +2ms
  docker-exec-server got stream end +0ms
  modem Sending: {
  path: '/exec/99097b8eeb63248a24956714c9cae2694ed08fde0f6bfdf4a092f60c59757f00/json',
  method: 'GET',
  headers: {},
  key: undefined,
  cert: undefined,
  ca: undefined,
  socketPath: '/var/run/docker.sock'
} +17ms
  modem Received: {"ID":"99097b8eeb63248a24956714c9cae2694ed08fde0f6bfdf4a092f60c59757f00","Running":true,"ExitCode":null,"ProcessConfig":{"tty":false,"entrypoint":"wc","arguments":["-c"],"privileged":false},"OpenStdin":true,"OpenStderr":true,"OpenStdout":true,"CanRemove":false,"ContainerID":"c991255b41376381b9a42ea939019d647aaea2d35b6da73422c3e2854abb57ce","DetachKeys":"","Pid":0}
 +2ms
  docker-exec-server exit code: null +3ms
  docker-exec-server:data <Buffer c8 00> +6ms
  docker-exec-server 0 sessions remain +0ms
  docker-exec-websocket-server:lib:rcv <Buffer c8 00> +5ms
Exit code: 0
stdout: ''
stderr: ''
    ✓ docker exec wc -c (41ms)
  docker-exec-server client close +2ms
  docker-exec-server client close +0ms
  modem Sending: {
  path: '/containers/c991255b41376381b9a42ea939019d647aaea2d35b6da73422c3e2854abb57ce?v=true&force=true',
  method: 'DELETE',
  headers: {},
  key: undefined,
  cert: undefined,
  ca: undefined,
  socketPath: '/var/run/docker.sock'
} +3ms
  modem Received:  +672ms
</pre>
</details>

<summary>A failed run, v16.0.0</summary>
<details>
<pre>
  docker-exec-server websocket server created for path: "/a" +0ms
  docker-exec-server websocket server created for path: "/a" +0ms
  docker-exec-websocket-server:lib:client ws://localhost:60171/a?tty=false&command=wc&command=-c +0ms
  docker-exec-server connection received +5ms
  modem Sending: {
  path: '/containers/500004f41b499c3670027915d4b63e46b4eb9f6249255e7f513614ca0b6ebefa/exec',
  method: 'POST',
  headers: { 'Content-Type': 'application/json', 'Content-Length': 105 },
  key: undefined,
  cert: undefined,
  ca: undefined,
  socketPath: '/var/run/docker.sock'
} +8ms
  docker-exec-server 1 sessions created +1ms
  docker-exec-websocket-server:lib:client socket opened +6ms
  docker-exec-websocket-server:lib:client client connected +0ms
  modem Received: {"Id":"fb9c043ca767fad5ecbe4cf1999be2d30136db30fb4e4e43b75a065ae2f650dc"}
 +14ms
  modem Sending: {
  path: '/exec/fb9c043ca767fad5ecbe4cf1999be2d30136db30fb4e4e43b75a065ae2f650dc/start',
  method: 'POST',
  headers: { 'Content-Type': 'application/json', 'Content-Length': 56 },
  key: undefined,
  cert: undefined,
  ca: undefined,
  socketPath: '/var/run/docker.sock'
} +1ms
  docker-exec-server:data <Buffer 64> +0ms
  docker-exec-server server executed session +30ms
  docker-exec-websocket-server:lib:rcv <Buffer 64> +0ms
  docker-exec-websocket-server:lib:client resumed +29ms
  docker-exec-websocket-server:lib:client 4 +0ms
  docker-exec-websocket-server:lib:client resumed +0ms
  docker-exec-websocket-server:lib:client 5 +0ms
  docker-exec-websocket-server:lib:client resumed +1ms
  docker-exec-websocket-server:lib:client 1 +0ms
  docker-exec-websocket-server:lib:client 0 +0ms
*** Long pause here ***
    1) docker exec wc -c
  docker-exec-server:data <Buffer c9> +40s
  docker-exec-server 0 sessions remain +40s
  modem Sending: {
  path: '/containers/500004f41b499c3670027915d4b63e46b4eb9f6249255e7f513614ca0b6ebefa?v=true&force=true',
  method: 'DELETE',
  headers: {},
  key: undefined,
  cert: undefined,
  ca: undefined,
  socketPath: '/var/run/docker.sock'
} +40s
  docker-exec-server client close +8ms
  docker-exec-server client close +0ms
  modem Received:  +804ms
</pre>
</details>

The key difference seems to be that the ``ExecStream`` never sees a 'end' event (I added some additional debug to detect this):

https://github.com/taskcluster/docker-exec-websocket-server/blob/b6ad23c13952e862210ce4659c70e819ef0d19e1/src/server.js#L148-L151

I'll look at the Node.js v16.0.0 changes and see if anything jumps out. I'd expect the change in buffers, streams, or events, but I may need to dig into [dockerode](https://github.com/apocas/dockerode) code to see what it is using.

**Comment by jwhitlock** [comment]
I decided to try upgrading dependencies first.

I've tested updating all dependencies with `yarn upgrade`, in the client and the server. I've packed up the client with `npm pack`, and installed that in the server (so that the in-docker build will work as well). The test breaks the same way.

There were some major versions remaining, some which update the node support:

```
Package  Current Wanted Latest Package Type    URL                                     
eslint   6.8.0   6.8.0  8.8.0  devDependencies https://eslint.org                      
mocha    6.2.3   6.2.3  9.2.0  devDependencies https://mochajs.org/                    
through2 3.0.2   3.0.2  4.0.2  dependencies    https://github.com/rvagg/through2#readme
ws       7.5.6   7.5.6  8.4.2  dependencies    https://github.com/websockets/ws   
```

I ran ``yarn upgrade --latest`` to update all of these, and bumped the client version as well. `eslint 8.8.0` failed in the v15.14.0 version, so I had to change to ``yarn install --ignore-engines``. And then, both tests failed in "before all" hook.

I'll step back and upgrade more slowly to figure out which package broken the tests.

**Comment by jwhitlock** [comment]
I was able to update:
* eslint - config file changes, replace some deprecated rules, drop Node 8 and 10, and apply some formatting
* mocha - Move config from `mocha.opts` to `package.json`. The configuration includes extending the timeout, which may be what broke the tests previously.
* through2 - No changes

Updating `ws` triggered some issues on my machine, even when reverting the update. I think this is a sign to stop for the day and reboot my laptop.

(Update - I wiped out `node_modules` and reinstalled, and that seemed to help.)

**Comment by jwhitlock** [comment]
The fix was in PR #35:

> Use [hijacking](https://docs.docker.com/engine/api/v1.22/#32-hijacking), a feature introduced in Docker engine v1.22, released with Docker 1.10 (February 2016). This allows the socket connection to better support stdin, stdout, and stderr streams, including closing a non-tty stdin stream (also see [dockerode: Streams goodness](https://github.com/apocas/dockerode#streams-goodness)).

This was broken in node v14, resulting in an empty stdout but a correctly reported exit code. In node v16, the "end" event was also blocked, causing the process to hang and getting some more attention for this code path.
